src/Core/src/Stream/ResponseBodyStream.php

```diff
@@ -33,45 +44,47 @@ public function __toString()
      */
     public function getChunks(): iterable
     {
+        if (null !== $this->fallback) {
+            return $this->fallback->getChunks();
+        }
+
+        if (null === $this->resource) {
+            $this->resource = \fopen('php://temp', 'rb+');
+        }
         foreach ($this->responseStream as $chunk) {
-            yield $chunk->getContent();
+            $chunkContent = $chunk->getContent();
+            \fwrite($this->resource, $chunkContent);
+            yield $chunkContent;
         }
+
+        $this->fallback = new ResponseBodyResourceStream($this->resource);
     }
 
     /**
      * {@inheritdoc}
      */
     public function getContentAsString(): string
     {
-        $resource = $this->getContentAsResource();
-
-        try {
-            return \stream_get_contents($resource);
-        } finally {
-            \fclose($resource);
+        if (null === $this->fallback) {
+            // Use getChunks() to read stream content to $this->fallback
+            foreach ($this->getChunks() as $chunk) {
+            }
         }
+
+        return $this->fallback->getContentAsString();
     }
 
     /**
      * {@inheritdoc}
      */
     public function getContentAsResource()
     {
-        $resource = \fopen('php://temp', 'rw+');
-
-        try {
-            foreach ($this->responseStream as $chunk) {
-                fwrite($resource, $chunk->getContent());
+        if (null === $this->fallback) {
+            // Use getChunks() to read stream content to $this->fallback
+            foreach ($this->getChunks() as $chunk) {
```

Comment:
That's a dummy implementation. A better solution would be to register a stream wrapper, and read the chunk ONLY when the stream is consumed